### PR TITLE
`PlusEqualProduct` version of `GFMultiplication` for GF($2^m$)

### PR DESCRIPTION
**Pull request by tanujkhattar**
Updates the existing `GFMultiplication` bloq to support an in-place `PlusEqualProduct` version of the bloq where the multiplied integer gets added to the result register. The decomposition is _very_ similar except we need an inverse of the linear reversible circuit at the beginning, so I decided to add this as a special case controlled by a flag instead of adding a new Bloq. 

Part of a larger effort to support galois field arithmetic in qualtran. xref https://github.com/quantumlib/Qualtran/issues/1419

### CODE REVIEW

**Review by NoureldinYosri — APPROVED** [review]
LGTM%important comment

**Comment by mpharrigan** [comment]
looks fine to me, modulo nour's comment. If that is indeed an issue, can we write a unit test that would have caught it?

File: qualtran/bloqs/gf_arithmetic/gf2_multiplication.py [hunk 1]
```diff
@@ -143,14 +156,15 @@ def qgf(self) -> QGF:
     def reduction_matrix_q(self) -> np.ndarray:
         m = int(self.bitsize)
         f = self.qgf.gf_type.irreducible_poly
-        M = np.zeros((m - 1, m))
+        M = np.zeros((m, m))
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/gf_arithmetic/gf2_multiplication.py:159]: No, the original matrix represents a transformation that maps m - 1 inputs to m outputs. The bloq for synthesizing the circuit assumes it acts on m inputs and outputs (notice that the signature has only one THRU register of size m) and the m'th input is mapped to the output as an identity. This change just makes the implicit assumption explicit, and it now has the benefit that the matrix is square and invertible so it works for both the cases. 

Tl;Dr - It's working as intended.
